Leanplum-SDK/Classes/LeanplumRequest.m

```diff
@@ -363,25 +367,49 @@ - (void)sendRequests:(BOOL)async
                     }
                 }
             }
-            
-            // Delete events on success.
-            [LPEventDataManager deleteEventsWithLimit:requestsToSend.count];
-            
-            // Send another request if the last request had maximum events per api call.
-            if (requestsToSend.count == MAX_EVENTS_PER_API_CALL) {
-                [self sendRequests:async];
-            }
-            dispatch_semaphore_signal(semaphore);
             LP_END_TRY
             
-            if (_eventIndex >= requestsToSend.count) {
-               _eventIndex -= requestsToSend.count;
-            } else if (_response) {
-                id response = [LPResponse getResponseAt:_eventIndex fromDictionary:json];
-                _response(operation, response);
+            // We need to lock sendNowCallbackMap so that new event callback won't be triggered
+            // right after it gets deleted.
+            NSMutableDictionary *callbackMap = [LeanplumRequest sendNowCallbackMap];
+            @synchronized (callbackMap) {
+                LP_TRY
+                // Delete events on success.
+                [LPEventDataManager deleteEventsWithLimit:requestsToSend.count];
+                
+                // Send another request if the last request had maximum events per api call.
+                if (requestsToSend.count == MAX_EVENTS_PER_API_CALL) {
+                    [self sendRequests:async];
+                }
+                dispatch_semaphore_signal(semaphore);
+                LP_END_TRY
+                
+                // Loop through all the possible callbacks.
+                // Note we need to do this because Request can steal future sendNow callbacks.
+                // Callback map will either have to be updated or removed.
+                NSMutableDictionary *updatedCallbackMap = [NSMutableDictionary new];
+                for (NSNumber *eventIndexObject in callbackMap.allKeys) {
+                    NSInteger eventIndex = [eventIndexObject integerValue];
```

Comment:
NSDictionary can only take specific NSObject as key/object. Unfortunately integer is not an object so we always have to convert to NSNumber to use properly. I was also not able to do what anna proposed (using LeanplumRequest or callback block as key) because of the same reason.